### PR TITLE
3/3 Feat connector secret

### PR DESCRIPTION
**Pull request by Sijoma**
Create connector secrets.

Since the update API is not documented and I dont want to mess with the POST endpoint for now. Updating a secret value requires a replacement of the resource. I am not sure if its even a good idea to allow to "view" the secret again / change it. But maybe its easier this way.

I would be interested whether my changes for checking whether a secret has been modified in the `READ` call can be improved. I just wrote it this way and tested it out without checking if thats the correct approach. Relevant lines start at `internal/provider/camunda_cluster_connector_secret_resource.go:171`

### CODE REVIEW

**Comment by multani** [comment]
Can you merge the change from `main`?

**Comment by Sijoma** [comment]
Did a rebase and had some conflicts in the docs :D. But should be okay.

**Comment by Sijoma** [comment]
Golangci-lint has some issues in GHA it seems - its probably caching.

```
   Downloading https://github.com/golangci/golangci-lint/releases/download/v1.52.2/golangci-lint-1.52.2-linux-amd64.tar.gz ...
  /usr/bin/tar xz --overwrite --warning=no-unknown-keyword --overwrite -C /home/runner -f /home/runner/work/_temp/755a5941-ed80-4c3c-a5e9-f3484f5e5282
  Installed golangci-lint into /home/runner/golangci-lint-1.52.2-linux-amd64/golangci-lint in 673ms
  Received 113705419 of 113705419 (100.0%), 129.6 MBs/sec
  Cache Size: ~108 MB (113705419 B)
  /usr/bin/tar -xf /home/runner/work/_temp/ee16c8ec-9494-4483-83be-63345f7cb65f/cache.tgz -P -C /home/runner/work/terraform-provider-camunda/terraform-provider-camunda -z
  Error: /usr/bin/tar: ../../../go/pkg/mod/google.golang.org/appengine@v1.6.7/appengine_vm.go: Cannot open: File exists
  /usr/bin/tar: ../../../go/pkg/mod/google.golang.org/appengine@v1.6.7/LICENSE: Cannot open: File exists
  Error: /usr/bin/tar: ../../../go/pkg/mod/google.golang.org/appengine@v1.6.7/mail/mail_test.go: Cannot open: File exists
  Error: /usr/bin/tar: ../../../go/pkg/mod/google.golang.org/appengine@v1.6.7/mail/mail.go: Cannot open: File exists
  /usr/bin/tar: ../../../go/pkg/mod/google.golang.org/appengine@v1.6.7/travis_install.sh: Cannot open: File exists
  Error: /usr/bin/tar: ../../../go/pkg/mod/google.golang.org/appengine@v1.6.7/cmd/aedeploy/aedeploy.go: Cannot open: File exists
  Error: /usr/bin/tar: ../../../go/pkg/mod/google.golang.org/appengine@v1.6.7/cmd/aebundler/aebundler.go: Cannot open: File exists
```

**Comment by Sijoma** [comment]
Tried to rerun linting now a couple of times. Not sure why its not working.

File: internal/provider/camunda_cluster_connector_secret_resource.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	ctx = context.WithValue(ctx, console.ContextAccessToken, r.provider.accessToken)
+
+	if data.Name == types.StringValue("") {
+
+	}
```
Comment (multani) [inline internal/provider/camunda_cluster_connector_secret_resource.go:158]: Actually, golangci works "correctly" and it reports an empty branch here, can you take a look?

Comment (Sijoma) [reply internal/provider/camunda_cluster_connector_secret_resource.go:158]: I meant the Github action (its failing with errors) and I cannot see the linting (scrolled over the laggy text).

I run it now locally and check the errors.

Comment (Sijoma) [reply internal/provider/camunda_cluster_connector_secret_resource.go]: Ahh! Shoot - my browser was always lagging and I have not seen this! Nice ❤️

**Comment by Sijoma** [comment]
Actually fixed that "name" and "key" were redundant. 

Console names the "key" of the secret in the api "name". On the frontend its displayed only as "key" and "value". It does not have a dedicated name.

image

File: internal/provider/camunda_cluster_connector_secret_resource.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	if req.ProviderData == nil {
+		return
+	}
+
+	provider, ok := req.ProviderData.(*CamundaCloudProvider)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *CamundaCloud client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
```
Comment (multani) [inline internal/provider/camunda_cluster_connector_secret_resource.go:82]:
```suggestion
			fmt.Sprintf("Expected *CamundaCloudProvider, got: %T. Please report this issue to the provider developers.", req.ProviderData),
```

File: internal/provider/camunda_cluster_connector_secret_resource.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+			fmt.Sprintf("Unable to create cluster connector secret, got error: %s",
+				err.(*console.GenericOpenAPIError).Body()),
+		)
+		return
+	}
+
+	if inline.StatusCode >= 300 {
+		resp.Diagnostics.AddError(
+			"Unable to create cluster connector secret",
+			fmt.Sprintf("Unable to create cluster connector secret, got status: %s", inline.Status))
```
Comment (multani) [inline internal/provider/camunda_cluster_connector_secret_resource.go:125]: `inline` is a bit of a weird variable name here perhaps?

```suggestion
	response, err := r.provider.client.ClustersApi.
		CreateSecret(ctx, data.ClusterId.ValueString()).
		CreateSecretBody(newClusterConnectorSecretConfiguration).
		Execute()

	if err != nil {
		resp.Diagnostics.AddError(
			"Unable to create cluster connector secret",
			fmt.Sprintf("Unable to create cluster connector secret, got error: %s",
				err.(*console.GenericOpenAPIError).Body()),
		)
		return
	}

	if response.StatusCode >= 300 {
		resp.Diagnostics.AddError(
			"Unable to create cluster connector secret",
			fmt.Sprintf("Unable to create cluster connector secret, got status: %s", response.Status))
```

File: internal/provider/camunda_cluster_connector_secret_resource.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+	// Find secrets name in list of secrets
+	// We also check if the value changed
+	found := false
+	for key, value := range secrets {
+		if key == data.Name.ValueString() {
+			found = true
+			if value != data.Value.ValueString() {
+				data.Value = types.StringValue(value)
+			}
```
Comment (multani) [inline internal/provider/camunda_cluster_connector_secret_resource.go:168]: You can probably remove the test here, not sure it's super useful:

```suggestion
			data.Value = types.StringValue(value)
```

**Review by multani — CHANGES_REQUESTED** [review]

**Review by multani — APPROVED** [review]
:rocket: thanks for the patch!

**Pull request merged**